lib/Pass/Pass.cpp

```diff
@@ -271,6 +271,42 @@ struct SouperPass : public ModulePass {
         return Builder.CreateAShr(V0, V1);
       case Inst::LShr:
         return Builder.CreateLShr(V0, V1);
+      case Inst::Ne:
+        if (isa<Constant>(V0)) {
+          return Builder.CreateICmpNE(V1, V0);
+        } else {
+          return Builder.CreateICmpNE(V0, V1);
+        }
+      case Inst::Eq:
+        if (isa<Constant>(V0)) {
+          return Builder.CreateICmpEQ(V1, V0);
+        } else {
+          return Builder.CreateICmpEQ(V0, V1);
+        }
+      case Inst::Ult:
+        if (isa<Constant>(V0)) {
+          return Builder.CreateICmpUGE(V1, V0);
+        } else {
+          return Builder.CreateICmpULT(V0, V1);
```

Comment:
not the right instruction!

Reply:
it would have been better for you to catch this yourself using a test case!
please fix this and review the rest and then I'll look again

Reply:
I meant UGE isn't right, obviously ULT is fine :)

Reply:
fixed